k9mail/src/main/java/com/fsck/k9/controller/ImapSyncInteractor.java

```diff
@@ -0,0 +1,268 @@
+package com.fsck.k9.controller;
+
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Set;
+
+import com.fsck.k9.Account;
+import com.fsck.k9.Account.Expunge;
+import com.fsck.k9.K9;
+import com.fsck.k9.mail.AuthenticationFailedException;
+import com.fsck.k9.mail.Folder;
+import com.fsck.k9.mail.Message;
+import com.fsck.k9.mail.MessagingException;
+import com.fsck.k9.mail.Store;
+import com.fsck.k9.mail.store.imap.ImapFolder;
+import com.fsck.k9.mail.store.imap.QresyncResponse;
+import com.fsck.k9.mailstore.LocalFolder;
+import com.fsck.k9.mailstore.LocalFolder.MoreMessages;
+import com.fsck.k9.mailstore.LocalMessage;
+import timber.log.Timber;
+
+
+class ImapSyncInteractor {
+
+    private final Account account;
+    private final String folderName;
+    private LocalFolder localFolder;
+    private ImapFolder imapFolder;
+    private final  MessagingListener listener;
+    private final  MessagingController controller;
+
+    ImapSyncInteractor(Account account, String folderName, MessagingListener listener, MessagingController controller) {
+        this.account = account;
+        this.folderName = folderName;
+        this.listener = listener;
+        this.controller = controller;
+    }
+
+    void performSync(MessageDownloader messageDownloader) {
+        Exception commandException = null;
+
+        try {
+            Timber.d("SYNC: About to process pending commands for account %s", account.getDescription());
+
+            try {
+                controller.processPendingCommandsSynchronous(account);
+            } catch (Exception e) {
+                controller.addErrorMessage(account, null, e);
+                Timber.e(e, "Failure processing command, but allow message sync attempt");
+                commandException = e;
+            }
+
+            Timber.v("SYNC: About to get local folder %s and open it", folderName);
+            localFolder = SyncUtils.getOpenedLocalFolder(account, folderName);
+            localFolder.updateLastUid();
+
+            Store remoteStore = account.getRemoteStore();
+            Timber.v("SYNC: About to get remote folder %s", folderName);
+            Folder remoteFolder = remoteStore.getFolder(folderName);
+            if (!(remoteFolder instanceof ImapFolder)) {
+                throw new MessagingException("A non-IMAP account was provided to ImapSyncInteractor");
```

Comment:
this seems like it's a bug if it ever happens, i.e. should crash with IllegalArgumentException